session/payment/session_balance.go

```diff
@@ -126,14 +135,73 @@ func (sb *SessionBalance) Start() error {
 func (sb *SessionBalance) loadInitialPromiseState() (promise.StoredPromise, error) {
 	lastPromise, err := sb.promiseStorage.GetLastPromise(sb.issuer)
 	if err != nil {
-		// if an error occurs when fetching the last promise, issue a new id
-		sb.sequenceID, err = sb.promiseStorage.GetNewSeqIDForIssuer(sb.issuer)
+		if err.Error() == errBoltNotFound.Error() {
+			// if not found, issue a new sequenceID
+			lastPromise.SequenceID, err = sb.promiseStorage.GetNewSeqIDForIssuer(sb.consumer, sb.receiver, sb.issuer)
+			sb.sequenceID = lastPromise.SequenceID
+			return lastPromise, err
+		}
 		return lastPromise, err
 	}
+
+	/* If we're providing a service for multiple consumers under the same issuer,
+	we'll need to check if the consumer ID matches.
+
+	If it does not - we'll need to do a thorough lookup since there's a case here,
+	where we might have issued a tiny amount that the provider did not clear,
+	but did clear the following promises.
+
+	In this case we'll need to issue a new id.
+	If none of the promises were cleared, we can reuse the old one.
+
+	TODO: This could turn out to be expensive, and we might just want to skip this check and issue a new ID instead.
+	*/
+	if lastPromise.ConsumerID != sb.consumer {
+		consumerPromise, err := sb.findPromiseForConsumer()
+		if err != nil {
+			if err.Error() == errNoPromiseForConsumer.Error() {
+				lastPromise.SequenceID, err = sb.promiseStorage.GetNewSeqIDForIssuer(sb.consumer, sb.receiver, sb.issuer)
+				sb.sequenceID = lastPromise.SequenceID
+				return lastPromise, err
+			}
+			return lastPromise, err
+		}
+		sb.sequenceID = consumerPromise.SequenceID
+		return consumerPromise, nil
+	}
+
 	sb.sequenceID = lastPromise.SequenceID
 	return lastPromise, nil
 }
 
+var errNoPromiseForConsumer = errors.New("no promise for consumer")
+
+func (sb *SessionBalance) findPromiseForConsumer() (promise.StoredPromise, error) {
```

Comment:
Feels like sorting and filtering helpers should be on `PromiseStorage` side

Reply:
moved